File: picasso-extensions/src/main/java/io/github/hendraanggrian/picasso/transformation/CropCircleTransformer.java
package io.github.hendraanggrian.picasso.transformation;

import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.support.annotation.NonNull;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class CropCircleTransformer extends CropSquareTransformer {

    @NonNull
    @Override
    public Bitmap transform(@NonNull Bitmap source, boolean recycleSource) {
        float r = Math.min(source.getWidth(), source.getHeight()) / 2f;
        final Bitmap squared = super.transform(source, recycleSource);
        final Bitmap circle = createDefaultBitmap(source);
        new Canvas(circle).drawCircle(r, r, r, new PaintBuilder(Paint.ANTI_ALIAS_FLAG)
                .shader(new BitmapShader(squared, BitmapShader.TileMode.CLAMP, BitmapShader.TileMode.CLAMP))
                .build());
        if (recycleSource)
            squared.recycle();
        return circle;
    }

    @Override
    public String key() {
        return Key.newInstance(this).toString();
    }
}

File: picasso-utils/src/com/hendraanggrian/picasso/transformation/CropCircleTransformer.kt
package com.hendraanggrian.picasso.transformation

import android.graphics.*
import com.squareup.picasso.Transformation

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
internal class CropCircleTransformer : Transformation {

    override fun transform(source: Bitmap): Bitmap {
        val size = Math.min(source.width, source.height)
        val width = (source.width - size) / 2
        val height = (source.height - size) / 2
        val target = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        val r = size / 2f
        Canvas(target).drawCircle(r, r, r, Paint().apply {
            val shader = BitmapShader(source, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
            if (width != 0 || height != 0) {
                // source isn't square, move viewport to center
                val matrix = Matrix()
                matrix.setTranslate(-width.toFloat(), -height.toFloat())
                shader.setLocalMatrix(matrix)
            }
            this.shader = shader
            this.isAntiAlias = true
        })
        source.recycle()
        return target
    }

    override fun key() = "CropCircleTransformer"
}